File: src/main/java/palindromeLinkedList/PalindromeLinkedList.java
/*
 * @Source: https://leetcode-cn.com/problems/palindrome-linked-list/
 * @Author: Tian Zhuo
 * @Date: 2019-01-18 22:23:05
 */
package palindromeLinkedList;

public class PalindromeLinkedList {
    public boolean isPalindrome(ListNode head) {
        if(head == null || head.next == null){
            return true;
        }
        ListNode fast = head;
        ListNode slow = head;
        while(fast.next != null && fast.next.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        slow = reverse(slow.next);
        while(slow != null){
            if(head.val != slow.val){
                return false;
            }
            head = head.next;
            slow = slow.next;
        }
        return true;
    }

    private ListNode reverse(ListNode head){
        if(head.next == null){
            return head;
        }
        ListNode newHead = reverse(head.next);
        head.next.next = head;
        head.next = null;
        return newHead;
    }


    public boolean isPalindrome1(ListNode head) {
        if (head == null || head.next == null) {
            return true;
        }
        if (head.next.next == null) {
            return head.val == head.next.val;
        }
        ListNode fast = head.next;
        ListNode slow = head;
        while (fast.next != null) {
            // 不停的从slow的后一个开始遍历，知道找到值相同的节点
            // 一次完成后，再移动到原节点的下一个节点开始，继续重复上面的步骤
            if (fast.next.val == slow.val) {
                if (fast.next.next != null) {
                    return false;
                }
                fast.next = null;
                slow = slow.next;
                fast = slow.next;
                if (fast == null || fast.val == slow.val) {
                    return true;
                }
            } else {
                fast = fast.next;
            }
        }
        return false;
    }
}

File: algorithms/js/palindromeLinkedList/palindromeLinkedList.js
/*
 * @Source: https://leetcode-cn.com/problems/palindrome-linked-list/
 * @Author: Tian Zhuo
 * @Date: 2019-01-19 11:25:16
 */
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    if(!head || !head.next) {
        return true;
    }
    let fast = head;
    let slow = head;
    while(fast.next && fast.next.next){
        fast = fast.next.next;
        slow = slow.next;
    }
    slow = reverse(slow.next);
    while(slow){
        if(head.val !== slow.val){
            return false;
        }
        slow = slow.next;
        head = head.next;
    }
    return true;

};
var reverse = function(head) {
    if(!head) {
        return null;
    }
    if(!head.next){
        return head;
    }
    let newHead = reverse(head.next);
    head.next.next = head;
    head.next = null;
    return newHead;
}

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome1 = function(head) {
    if(!head) return true;
    let s = head;
    let f = head;

    while(f && f.next && f.next.next) {
        s = s.next;
        f = f.next.next;
    }
    let right = s.next;
    s.next = null;

    let current = right;
    let last = null;
    while(current){
        const next = current.next;
        current.next = last;
        last = current;
        current = next;
    }

    a = head;
    b = last;

    while(a && b) {
        if(a.val !== b.val) {
            return false;
        }
        a = a.next;
        b = b.next;
    }
    return true;
};